r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs

```diff
@@ -778,84 +833,164 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
-            limbs: prod_limbs,
-            num_add_over_normal_form: num_add_bound - BigUint::one(),
+        let result = NonNativeFieldGadget {
+            limbs: reduced_prod_limbs,
+            num_of_additions_over_normal_form: &num_add_bound - &BigUint::one(),
             simulation_phantom: PhantomData,
-        })
+        };
+
+        debug_assert!(
+            result.check(),
+            "mul_without_prereduce_for_pseudomersenne(): result failed on check()"
+        );
+
+        Ok(result)
     }
 
-    /// Enforces two non-native gadgets, not necessarily in normal form, to be equal mod the 
-    /// non-native modulus `p`. Assumes that 
+    /// For advanced use, multiply and output the intermediate representations (without reduction)
+    /// This intermediate representations can be added with each other, and they can later be
+    /// reduced back to the `NonNativeFieldGadget`.
+    /// Assumes that
+    /// ``
+    ///     2 * bits_per_limb + surfeit' <= CAPACITY - 2,
+    /// ``
+    /// where
+    /// ``
+    ///      surfeit' = log(num_limbs^2 * (num_add(L)+1) + 1).
+    /// ``
+    pub fn mul_by_constant_without_prereduce<CS: ConstraintSystemAbstract<ConstraintF>>(
+        &self,
+        mut cs: CS,
+        other: &SimulationF,
+    ) -> Result<NonNativeFieldMulResultGadget<SimulationF, ConstraintF>, SynthesisError> {
+        let other_gadget =
+            NonNativeFieldGadget::from_value(cs.ns(|| "alloc constant for mul"), other);
+        // Check if SimulationF is pseudo-mersenne, as in this case we call the optimized
+        // multiplication function for pseudo-mersenne fields
+        return if super::is_pseudo_mersenne::<SimulationF>() {
+            let res = self.mul_without_prereduce_for_pseudomersenne(
+                cs.ns(|| "mul for pseudo-mersenne"),
+                &other_gadget,
+                true,
+            )?;
+            // convert res to NonNativeFieldMulResultGadget to be compatible with type returned by the function
+            FromGadget::from(&res, &mut cs)
+        } else {
+            self.mul_without_prereduce_no_pseudomersenne(
+                cs.ns(|| "mul no pseudo-mersenne"),
+                &other_gadget,
+                true,
+            )
+        };
+    }
+
+    /// Enforces two non-native gadgets, not necessarily in normal form, to be equal mod the
+    /// non-native modulus `p`. Assumes that
     /// ``
     ///    bits_per_limb + surfeit <= CAPACITY - 2,
     /// ``
-    /// where 
+    /// where
     /// ``
     ///    surfeit = 1 + log(3 + num_add(L) + num_add(R)).
     /// ``
-    // Costs 
+    // Costs
     // ``
-    //     C = 3 * S + surfeit + num_limbs(p) + 1,
+    //     C = (num_groups - 1) * (surfeit + 4) + 2 + surfeit + num_limbs(p) + 1,
     // ``
-    // where 
+    // where
     // ``
-    //      S =  2 + Floor[
+    //      S =  Floor[
     //          (ConstraintF::CAPACITY - 2 - surfeit) / bits_per_limb
```

Comment:
You have to call S "num_groups" according to the new notation.

Reply:
S is the number of limbs per group, I forgot to write that num_groups = Ceil[num_limbs/S], sorry. I will add it in the next commit.